File: semantic-efsm/src/main/java/efsm/semantic/algebra/EFSMAlgebra.java
package efsm.semantic.algebra;

import syntax.efsm.FSM;
import syntax.fsm.State;
import syntax.fsm.Transition;

/**
 * Created by mleduc on 13/03/17.
 */
public interface EFSMAlgebra<TFSM, TState, TTransition> {
    TFSM fsm(syntax.fsm.FSM fsm);
    TFSM fsm(FSM fsm);
    TState state(State state);
    TTransition transition(Transition transition);

    default TFSM $(syntax.fsm.FSM fsm) {
        if (fsm instanceof syntax.efsm.FSM) {
            return fsm((syntax.efsm.FSM) fsm);
        }
        return fsm(fsm);
    }

    default TState $(State state) {
        return state(state);
    }

    default TTransition $(Transition transition) {
        return transition(transition);
    }

}

File: semantic-efsm-scala/src/efsm/semantics/scla/EfsmAlgebra.scala
package efsm.semantics.scla

/**
  * Created by mleduc on 28/03/17.
  */
trait EfsmAlgebra[FSMT, eFSMT <: FSMT, TransitionT, StateT] {
  def fsm(fsm: syntax.efsm.FSM): eFSMT

  def fsm(fsm: syntax.fsm.FSM): FSMT

  def transition(transition: syntax.fsm.Transition): TransitionT

  def state(state: syntax.fsm.State): StateT
}
